src/lib/y2users/config.rb

```diff
@@ -48,31 +61,171 @@ def system(reader: nil, force_read: false)
+    end
+
+    # Attaches users and groups to this config
+    #
+    # The given users and groups cannot be already attached to a config.
+    #
+    # @param elements [Array<User, Group>]
+    def attach(*elements)
+      elements.each { |e| attach_element(e) }
+    end
+
+    # Detaches users and groups from this config
+    #
+    # @param elements [Array<User, Group>]
+    def detach(*elements)
+      elements.each { |e| detach_element(e) }
+    end
+
+    # Generates a new config with the very same list of users and groups
+    #
+    # Note that the cloned users and groups keep the same id as the original users and groups.
+    #
+    # @param name [String] name for the new cloned config
+    # @return [Config]
     def clone_as(name)
       config = self.class.new(name)
-      config.users = users.map { |u| u.clone_to(config) }
-      config.groups = groups.map { |g| g.clone_to(config) }
+
+      elements = users + groups
+      elements.each { |e| config.clone_element(e) }
 
       config
     end
+
+  protected
+
+    # Clones a given user or group and attaches it into this config
+    #
+    # Note that the cloned element keep the same id as the source element.
+    #
+    # @param element [User, Group]
+    def clone_element(element)
+      cloned = element.clone
+      cloned.assign_internal_id(element.id)
+
+      attach(cloned)
+    end
+
+  private
+
+    # Manager for users
+    #
+    # @return [ElementManager]
+    attr_reader :users_manager
+
+    # Manager for groups
+    #
+    # @return [ElementManager]
+    attr_reader :groups_manager
+
+    # Generates the id for the next attached user or group
+    #
+    # @return [Integer]
+    def self.next_element_id
+      @last_element_id ||= 0
+      @last_element_id += 1
+    end
+
+    private_class_method(:next_element_id)
+
+    # Attaches an user or group
+    #
+    # An id is assigned to the given user/group, if needed.
+    #
+    # @param element [User, Group]
+    def attach_element(element)
+      element.assign_internal_id(self.class.next_element_id) if element.id.nil?
+
+      element.is_a?(User) ? users_manager.attach(element) : groups_manager.attach(element)
+    end
+
+    # Detaches an user or group
+    #
+    # @param element [User, Group]
+    def detach_element(element)
+      element.is_a?(User) ? users_manager.detach(element) : groups_manager.detach(element)
+    end
+
+    # Helper class to manage a list of users or groups
+    class ElementManager
+      # @return [Array<User, Group>]
+      attr_reader :elements
+
+      # Constructor
+      #
+      # @param config [Config]
+      def initialize(config)
+        @config = config
+        @elements = []
+      end
+
+      # Attaches the element to the config
+      #
+      # @raise [RuntimeError] if the element is already attached
+      #
+      # @param element [User, Group]
+      def attach(element)
+        raise "Element already attached: #{element}" if element.attached?
```

Comment:
Suggestion: since the element can be attached to other config too, perhaps worth it to include _where_ (the config name?) is attached.

Reply:
I expect that is shown as part of `element.inspect`.

Reply:
Ah, right :sweat_smile: